client/core/core.go

```diff
@@ -1373,79 +1374,100 @@ func (c *Core) initializeDEXConnections(crypter encrypt.Crypter) []*DEXBrief {
 	// errorSet.
 	var wg sync.WaitGroup
 	c.connMtx.RLock()
-	defer c.connMtx.RUnlock()
 	results := make([]*DEXBrief, 0, len(c.conns))
-	for _, dc := range c.conns {
-		dc.tradeMtx.RLock()
-		tradeIDs := make([]string, 0, len(dc.trades))
-		for tradeID := range dc.trades {
-			tradeIDs = append(tradeIDs, tradeID.String())
-		}
-		dc.tradeMtx.RUnlock()
-		result := &DEXBrief{
-			Host:     dc.acct.host,
-			TradeIDs: tradeIDs,
-		}
-		results = append(results, result)
-		// Copy the iterator for use in the authDEX goroutine.
-		if dc.acct.authed() {
-			result.Authed = true
-			result.AcctID = dc.acct.ID().String()
-			continue
-		}
-		err := dc.acct.unlock(crypter)
-		if err != nil {
-			details := fmt.Sprintf("error unlocking account for %s: %v", dc.acct.host, err)
-			c.notify(newFeePaymentNote("Account unlock error", details, db.ErrorLevel, dc.acct.host))
-			result.AuthErr = details
-			continue
-		}
-		result.AcctID = dc.acct.ID().String()
-		dcrID, _ := dex.BipSymbolID("dcr")
-		if !dc.acct.feePaid() {
-			if len(dc.acct.feeCoin) == 0 {
-				details := fmt.Sprintf("Empty fee coin for %s.", dc.acct.host)
-				c.notify(newFeePaymentNote("Fee coin error", details, db.ErrorLevel, dc.acct.host))
-				result.AuthErr = details
-				continue
+	var enabledDEXConnections = make(map[string]*dexConnection)
+	for host, dc := range c.conns {
+		wg.Add(1)
+		go func(host string, dc *dexConnection) {
+			defer wg.Done()
+			dc.tradeMtx.RLock()
+			tradeIDs := make([]string, 0, len(dc.trades))
+			for tradeID := range dc.trades {
+				tradeIDs = append(tradeIDs, tradeID.String())
 			}
-			// Try to unlock the Decred wallet, which should run the reFee cycle, and
-			// in turn will run authDEX.
-			dcrWallet, err := c.connectedWallet(dcrID)
-			if err != nil {
-				log.Debugf("Failed to connect for reFee at %s with error: %v", dc.acct.host, err)
-				details := fmt.Sprintf("Incomplete registration detected for %s, but failed to connect to the Decred wallet", dc.acct.host)
-				c.notify(newFeePaymentNote("Wallet connection warning", details, db.WarningLevel, dc.acct.host))
-				result.AuthErr = details
-				continue
+			dc.tradeMtx.RUnlock()
+			result := &DEXBrief{
+				AcctID:   dc.acct.ID().String(),
+				Host:     dc.acct.host,
+				TradeIDs: tradeIDs,
+				Authed:   false,
 			}
-			if !dcrWallet.unlocked() {
-				err = unlockWallet(dcrWallet, crypter)
-				if err != nil {
-					details := fmt.Sprintf("Connected to Decred wallet to complete registration at %s, but failed to unlock: %v", dc.acct.host, err)
-					c.notify(newFeePaymentNote("Wallet unlock error", details, db.ErrorLevel, dc.acct.host))
-					result.AuthErr = details
-					continue
-				}
+			results = append(results, result)
```

Comment:
There's a good chance that I expanded the original goroutine responsibility beyond what it should be.

It looks like initializeDEXConnections was originally processing calls and responses to authDEX() in parallel. I'm assuming it was done in parallel because the authDEX() calls can take time waiting for a response. Would you say as a general rule that service calls from within a loop should always be done in parallel?

The order of operations for initializeDEXConnections() was changed because I noticed that there was no reason to continue processing the dex connection if the user could not be authenticated.

With the assumption that we still prefer to execute authDEX() in parallel I will look for opportunities to trim down what is happening within the goroutine.

Reply:
Your refactor has introduced a few bugs. In addition to the concurrent `result` writes, you are trying to auth connections before you check that they've paid. Please revert.

Reply:
@buck54321 Okay, what you're saying "you are trying to auth connections before you check that they've paid" does make sense. 

The following is how I arrived at my current version of `initializeDEXConnections()`.

The following statements are both true:

1. We do not need/want to attempt `authDEX()` if the account is not paid according to the client records.

2. We do not need to do any further login activity if the account is not found because you will have to re-register and concern yourself with paying another fee.

I think I went down path 2 because it did not make sense to tell the user that they need to pay the fee if we know they will have to register the dex again and pay another fee. Am I thinking about this correctly? Is it such an edge case that we should not concern ourselves? 

Since we have an issue written up to reinstate an account based on client records https://github.com/decred/dcrdex/issues/515, maybe we should continue looking for fee paid as the user will avoid repayment altogether. In that case we would indeed want to check client records for fee paid first. I think I've just convinced myself to do the revert as you suggest. I will go in that direction unless I hear otherwise.